### ISSUE TITLE
GetRecords Not Work on Xamarin.iOS

### ISSUE THREAD

**Issue by MrMYHuang** [JoshClose/CsvHelper#649]
I have a Xamarin project created by Visual Studio 2017 RC:
[https://github.com/MrMYHuang/tfwc/tree/720f036b866429cf2585a11ca0206e185ae08657]()
In the Git repo, the file tfwc\tfwc.Portable\tfwcTabPage.xaml.cs calls GetRecords (CsvHelpler ver. 2.16.3) in one line. It works well on Android and UWP (Windows 10), but unfortunately, not work on iOS (simulator) and I got its exception stack trace:
```
{System.NotImplementedException: The method or operation is not implemented.
  at Microsoft.Scripting.Interpreter.LightCompiler.CompileMemberInitExpression (System.Linq.Expressions.Expression expr) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/LightCompiler.cs:1532 
  at Microsoft.Scripting.Interpreter.LightCompiler.CompileNoLabelPush (System.Linq.Expressions.Expression expr) [0x002a7] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/LightCompiler.cs:1660 
  at Microsoft.Scripting.Interpreter.LightCompiler.Compile (System.Linq.Expressions.Expression expr) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/LightCompiler.cs:1729 
  at Microsoft.Scripting.Interpreter.LightCompiler.CompileTop (System.Linq.Expressions.LambdaExpression node) [0x00055] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/LightCompiler.cs:221 
  at Microsoft.Scripting.Generation.CompilerHelpers.LightCompile (System.Linq.Expressions.LambdaExpression lambda) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Generation/CompilerHelpers.cs:619 
  at System.Linq.Expressions.Compiler.LambdaCompiler.Compile (System.Linq.Expressions.LambdaExpression lambda, System.Runtime.CompilerServices.DebugInfoGenerator debugInfoGenerator) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/System.Core/System.Linq.Expressions.Interpret/LambdaCompiler.cs:41 
  at System.Linq.Expressions.LambdaExpression.Compile () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/10.4.0.123/src/mono/mcs/class/dlr/Runtime/Microsoft.Scripting.Core/Ast/LambdaExpression.cs:127 
  at CsvHelper.CsvReader.CreateFuncForObject (System.Type recordType) [0x000a6] in <cd3945f1299845aabbe2d9f53cc65e3f>:0 
  at CsvHelper.CsvReader.CreateReadRecordFunc (System.Type recordType) [0x00053] in <cd3945f1299845aabbe2d9f53cc65e3f>:0 
  at CsvHelper.CsvReader.GetReadRecordFunc[T] () [0x0000b] in <cd3945f1299845aabbe2d9f53cc65e3f>:0 
  at CsvHelper.CsvReader.CreateRecord[T] () [0x00000] in <cd3945f1299845aabbe2d9f53cc65e3f>:0 
  at CsvHelper.CsvReader+<GetRecords>d__65`1[T].MoveNext () [0x0002b] in <cd3945f1299845aabbe2d9f53cc65e3f>:0 }
```
I find a similar discussion here:
[https://forums.xamarin.com/discussion/comment/90430]()
This issue has existed since 2014. Unfortunately, Xamarin doesn't fix it. Alternatively, could CsvHelper works around this issue by rewriting some code of GetRecords?

**Comment by JoshClose** [comment]
The issue is that iOS doesn't allow compiling expression trees while running, if I remember correctly. That means only the 2.0 version of CsvHelper can be used on iOS. Expression trees didn't exist until 3.5 or something like that, so it's not available in the 2.0 version.

If iOS can do this now I would gladly make changes for it to work. To test this out, I think you can try to do `Expression.Lambda(...).Compile()` which will create a delegate, then execute the delegate.

**Comment by MrMYHuang** [comment]
Thanks for your information.
Because formats of my CSV files are simple, I work around this issue by writing C# code myself.
